pkg/apis/wsecurity/v1alpha1/httpReq.go

```diff
@@ -0,0 +1,207 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+)
+
+//////////////////// ReqProfile ////////////////
+// Does not monitor Trailers
+
+// Exposes ValueProfile interface
+type ReqProfile struct {
+	ClientIp      IpSetProfile     `json:"cip"`           // 192.168.32.1
+	HopIp         IpSetProfile     `json:"hopip"`         // 1.2.3.4
+	Method        SetProfile       `json:"method"`        // GET
+	Proto         SetProfile       `json:"proto"`         // "HTTP/1.1"
+	MediaType     MediaTypeProfile `json:"mediatype"`     // "text/html"
+	ContentLength CountProfile     `json:"contentlength"` // 0
+	Url           UrlProfile       `json:"url"`
+	Qs            QueryProfile     `json:"qs"`
+	Headers       HeadersProfile   `json:"headers"`
+}
+
+func (profile *ReqProfile) profileI(args ...interface{}) {
+	profile.Profile(args[0].(*http.Request), args[1].(net.IP))
+}
+
+func (profile *ReqProfile) Profile(req *http.Request, cip net.IP) {
+	var hopIpStr string
+
+	profile.ClientIp.ProfileIP(cip)
+
+	// Future: Does not monitor rfc7239 "forwarded" headers
+	if forwarded, ok := req.Header["X-Forwarded-For"]; ok {
+		if numHops := len(forwarded); numHops > 0 {
+			hopIpStr = forwarded[numHops-1]
+		}
+	}
+	profile.HopIp.ProfileString(hopIpStr)
+
+	profile.MediaType.Profile(req.Header.Get("Content-Type"))
+	profile.Qs.Profile(req.URL.Query())
+
+	profile.Method.ProfileString(req.Method)
+	profile.Proto.ProfileString(req.Proto)
+	profile.Url.Profile(req.URL.Path)
+	profile.Headers.Profile(req.Header)
+
+	length := req.ContentLength
+	if length > 0 {
+		var log2length uint8
+
+		for length > 0 {
+			log2length++
+			length >>= 1
+		}
+		profile.ContentLength = CountProfile(log2length)
+	}
+}
+
+//////////////////// ReqPile ////////////////
+
+// Exposes ValuePile interface
+type ReqPile struct {
```

Comment:
Given that the `ReqProfile` and `ReqPile` types have the same elements, would there be any benefit to doing something like this:

```golang
type Req struct {
	ClientIp      IpSetProfile     `json:"cip"`           // 192.168.32.1
	HopIp         IpSetProfile     `json:"hopip"`         // 1.2.3.4
	Method        SetProfile       `json:"method"`        // GET
	Proto         SetProfile       `json:"proto"`         // "HTTP/1.1"
	MediaType     MediaTypeProfile `json:"mediatype"`     // "text/html"
	ContentLength CountProfile     `json:"contentlength"` // 0
	Url           UrlProfile       `json:"url"`
	Qs            QueryProfile     `json:"qs"`
	Headers       HeadersProfile   `json:"headers"`
}

type ReqProfile Req
type ReqPile Req
```

Reply:
(realize we didn't do it for previous types, so it might require a larger refactor)

Reply:
not the same elements...
while ReqProfile is a structure of IpSetProfile , SetProfile , UrlProfile etc. 
ReqPile is a structure of IpSetPile, SetPile, UrlPile etc.

Reply:
:facepalm: